### DISCUSSION TITLE
Request:差分を取り扱うことのできる変更下書きシステム

### DISCUSSION THREAD

**Discussion by strvert** [growilabs/growi#7705] [Ideas / Requests] [answered]
### 不便に感じている点
チームで検討しながら情報をアップデートしていく際、現状では別ページに変更内容の草案を作成し、会議検討の後に手動でコピペするなどして対応するしか無い。これは編集における使い勝手として理想的ではない。

### 提案機能
記事の変更を行う際、即座の上書きではなく、一旦変更を保持して後から差分を見ながら手動Merge編集が可能な機能がほしい。
VCS(Gitなど)が実現している機能を部分的に取り込むもの。

### 想定される機能の詳細
- 既に投稿されたページに対して情報の追記や変更を行う際、即座に上書きで反映してしまう以外に、下書きとして保持できる。
- 下書きページは独自のパーマリンクを保持しており、自分以外のユーザーにとっても閲覧可能なものである。
    - 下書きページは紐づく記事ページに対して複数保持可能である。
- 「下書きをMerge」などのボタンを押すことで、現状の記事内容に対するMergeが可能である。
    - Conflictがない場合、下書きの全体を反映する「一括Merge」が選択可能。
    - Conflictがある場合、あるいは話し合いなどの結果として部分的な変更のMergeが必要となった場合のために、「手動Merge」が選択可能。
- 「手動Merge」を選択すると、Mergeエディタが起動する。
    - Mergeエディタは、通常のGrowiのエディタに変更のある行への背景ハイライトを追加しただけのもので良いと考えられる。履歴の差分機能などからして、記事内容の差分を取得可能なエンジンは搭載済みであると思われるので、同様に利用することができるのではないか。参考としては、GitHubのConflict解消エディタや、VSCodeのMerge conflictsなど多くのエディタに搭載されているVCS機能が挙げられる。(以下参考)

> ![image](https://user-images.githubusercontent.com/20394825/120618861-6dba5a80-c496-11eb-9404-16d96d1e7943.png)
> GitHub

> ![image](https://user-images.githubusercontent.com/20394825/120618907-79a61c80-c496-11eb-973d-d8fb16e84704.png)
> VSCode

### 想定した代替案
- 全ページに対してVCSなどのようなシステムを適応し、個人ごとにブランチを作成することができるようにする。
    - メリット: 複数ページに対する編集をブランチ上での変更としてまとめることができるため、編集の意図に合わせて一括のPRなどが可能となる。
    - デメリット: Growiシステムへの変更が大きすぎる。また、VCSに慣れていない多くの一般ユーザーにとって直感的でない。

**Comment by yuki-takei** [comment]
@strvert こちら完全にご要望通りの機能ではありませんが、コンフリクトした際の解消機能の実装には着手し始めました。
https://github.com/weseek/growi/pull/4490

**Comment by yuki-takei** [comment, marked as answer]
[v4.5.6](https://github.com/weseek/growi/releases/tag/v4.5.6) で実装されました
